File: transfer.py
# Tracks the progress of transfers for the server because it needs to be able to track where its at 
#
# this is too complex to track the state of the client because when you run it ur only ever gonna 
# be making one request and be done once u receive what u asked for


# everytime the server is responding to a request with data, a Transfer object needs to be created to track the progress
class Transfer:

    def __init__(self, connection_id, client_addr, filename, file_data, segment_size, transfer_number):
        self.connection_id = connection_id
        self.client_addr = client_addr
        self.filename = filename
        self.file_data = file_data
        self.file_length = len(file_data)
        self.segment_size = segment_size
        self.is_complete = False 
        self.current_chunk = 0  #   goes from 0 to len(file_data)-1
        self.seq_num = 0 #   when we are transfering data, the first sequence number will always be zero. 
        self.transfer_number = transfer_number

        self.chunks = self.split_data(file_data)
        self.total_chunks = len(self.chunks)

        print(f"[SERVER]    Initiatied a Transfer object with {self.total_chunks} chunks of data. Chunk will be {segment_size} bytes")
        
    def split_data(self, data):
        """splits the file content into the appropriate number of chunks"""
        chunks = []
        for i in range(0, max(len(data), 1), self.segment_size):
            chunks.append(data[ i : i + self.segment_size])
        
        return chunks
    
    
    def get_current_chunk(self):
        """ Return the data of the chunk we are currently at"""
        return self.chunks[self.current_chunk]
    
    def next(self):
        """Increments the 'current_chunk' and 'seq_num' vars """
        self.current_chunk += 1
        self.seq_num = 1 - self.seq_num
    

File: java/src/transfer/Transfer.java
package transfer;

import java.net.InetSocketAddress;
import java.util.ArrayList;
import java.util.List;

/**
 * Tracks the state of a single file transfer on the server side.
 * The server can handle multiple concurrent transfers (one per connection ID),
 * so each active transfer is represented by one Transfer object.
 */
public class Transfer {

    public final long              connectionId;
    public final InetSocketAddress clientAddr;
    public final String            filename;
    public final byte[]            fileData;
    public final int               fileLength;
    public final int               segmentSize;
    public final int               transferNumber;

    public boolean isComplete    = false;
    public int     currentChunk  = 0;   // index into chunks list
    public int     seqNum        = 0;   // alternates 0/1 for stop-and-wait
    public int     retransmissions = 0;

    private final List<byte[]> chunks;
    public  final int          totalChunks;

    public Transfer(long connectionId, InetSocketAddress clientAddr,
                    String filename, byte[] fileData,
                    int segmentSize, int transferNumber) {
        this.connectionId   = connectionId;
        this.clientAddr     = clientAddr;
        this.filename       = filename;
        this.fileData       = fileData;
        this.fileLength     = fileData.length;
        this.segmentSize    = segmentSize;
        this.transferNumber = transferNumber;

        this.chunks      = splitData(fileData);
        this.totalChunks = chunks.size();

        System.out.printf("[SERVER]    Initiated a Transfer object with %d chunks of data. " +
                "Each chunk will be up to %d bytes%n", totalChunks, segmentSize);
    }

    /** Split file bytes into fixed-size chunks. */
    private List<byte[]> splitData(byte[] data) {
        List<byte[]> result = new ArrayList<>();
        int len = Math.max(data.length, 1);
        for (int i = 0; i < len; i += segmentSize) {
            int end = Math.min(i + segmentSize, data.length);
            byte[] chunk = new byte[end - i];
            System.arraycopy(data, i, chunk, 0, chunk.length);
            result.add(chunk);
        }
        // Edge case: empty file -> one empty chunk
        if (data.length == 0) {
            result.add(new byte[0]);
        }
        return result;
    }

    /** Return the chunk we are currently waiting to have acknowledged. */
    public byte[] getCurrentChunk() {
        return chunks.get(currentChunk);
    }

    /** Advance to the next chunk and flip the sequence number. */
    public void next() {
        currentChunk++;
        seqNum = 1 - seqNum;
    }
}
